tokio/src/net/tcp/split_owned.rs

```diff
@@ -0,0 +1,224 @@
+//! `TcpStream` owned split support.
+//!
+//! A `TcpStream` can be split into an `OwnedReadHalf` and a
+//! `OwnedWriteHalf` with the `TcpStream::split_owned` method.
+//! `OwnedReadHalf` implements `AsyncRead` while `OwnedWriteHalf`
+//! implements `AsyncWrite`.
+//!
+//! Compared to the generic split of `AsyncRead + AsyncWrite`, this specialized
+//! split has no associated overhead and enforces all invariants at the type
+//! level.
+
+use crate::future::poll_fn;
+use crate::io::{AsyncRead, AsyncWrite};
+use crate::net::TcpStream;
+
+use bytes::Buf;
+use std::error::Error;
+use std::mem::MaybeUninit;
+use std::net::Shutdown;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use std::{fmt, io};
+
+/// Owned read half of a [`TcpStream`], created by [`split_owned`].
+///
+/// [`TcpStream`]: TcpStream
+/// [`split_owned`]: TcpStream::split_owned()
+#[derive(Debug)]
+pub struct OwnedReadHalf(Arc<TcpStream>);
+
+/// Owned write half of a [`TcpStream`], created by [`split_owned`].
+///
+/// Note that in the `AsyncWrite` implemenation of this type, `poll_shutdown` will
+/// shut down the TCP stream in the write direction.
+///
+/// [`TcpStream`]: TcpStream
+/// [`split_owned`]: TcpStream::split_owned()
+#[derive(Debug)]
+pub struct OwnedWriteHalf(Arc<TcpStream>);
+
+pub(crate) fn split_owned(stream: TcpStream) -> (OwnedReadHalf, OwnedWriteHalf) {
+    let arc = Arc::new(stream);
+    let arc2 = Arc::clone(&arc);
+    (OwnedReadHalf(arc), OwnedWriteHalf(arc2))
+}
+
+pub(crate) fn reunite(
+    read: OwnedReadHalf,
+    write: OwnedWriteHalf,
+) -> Result<TcpStream, ReuniteError> {
+    if Arc::ptr_eq(&read.0, &write.0) {
+        drop(write);
+        // This unwrap cannot fail as the api does not allow creating more than two Arcs,
+        // and we just dropped the other half.
+        Ok(Arc::try_unwrap(read.0).expect("Too many handles to Arc"))
+    } else {
+        Err(ReuniteError(read, write))
+    }
+}
+
+/// Error indicating two halves were not from the same socket, and thus could
+/// not be reunited.
+#[derive(Debug)]
+pub struct ReuniteError(pub OwnedReadHalf, pub OwnedWriteHalf);
+
+impl fmt::Display for ReuniteError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "tried to reunite halves that are not from the same socket"
+        )
+    }
+}
+
+impl Error for ReuniteError {}
+
+impl OwnedReadHalf {
+    /// Attempts to put the two halves of a `TcpStream` back together and
+    /// recover the original socket. Succeeds only if the two halves
+    /// originated from the same call to [`split_owned`].
+    ///
+    /// [`split_owned`]: TcpStream::split_owned()
+    pub fn reunite(self, other: OwnedWriteHalf) -> Result<TcpStream, ReuniteError> {
```

Comment:
instead of having 2 `reunite` methods there could also be a `from_halves` method on the `TcpStream`.